02_bubble/CameronSchafer/bubble.go

```diff
@@ -0,0 +1,55 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	// requirement was to sort a copy of the array.
+	t := make([]int, len(s))
+	copy(t, s)
+	check := false
+
+	for !check {
+		swapped := true
+		// sort once and then check
+		for i := 0; i < len(t); i++ {
```

Comment:
Typically bubble sort's inner loop bound goes from 0 to i, with i being the out loop index ranging from n to 0.
After the first iteration of the outer loop the largest element in the slice has "bubbled up" to the last index and does not need to be considered for sorting any more (it already is sorted):
```go
	for i := len(s); i > 0; i-- {
		swapped := false
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
				swapped = true
			}
		}
		if !swapped {
			break
		}
	}
```
(feel free to use, adjust or ignore our reference solution).